### PR TITLE
impl `Map::retain` and `Set::retain`

### PR DESCRIPTION
**Pull request by pitaj**
also test examples and benches in CI

relates #11

### CODE REVIEW

File: examples/basic.rs [hunk 1]
```diff
@@ -12,3 +12,10 @@ fn main() {
     assert_eq!(map.get(Key::First), Some(&42));
     assert_eq!(map.get(Key::Second), None);
 }
+
+// Execute this during testing as well.
+#[cfg(test)]
+#[test]
+fn test_main() {
+    main();
+}
```
Comment (udoprog) [inline examples/basic.rs]: Examples don't need to be tested. They should be as compact as possible to showcase some basic or neat feature of the library. Think this is something people look at on GitHub to get a feel for the library. Less noise = better.

File: examples/map_feature.rs [hunk 2]
```diff
@@ -0,0 +1,47 @@
+// This file is very useful for debugging the derive macro.
+// Doctests don't work well for that purpose.
+
+fn main() {
+    #[cfg(feature = "map")]
```
Comment (udoprog) [inline examples/map_feature.rs]: Only build-test examples with the full feature set instead. Less noise = better.

**Review by udoprog — CHANGES_REQUESTED** [review]
Retain impl LGTM, I'd like the examples to be cleaner though. Easiest would be to only build examples with the full set of features (and no need to `#[test]` them).

**Comment by pitaj** [comment]
Understood on the examples, but can you restore the one you deleted (`map_feature`) and just remove the feature stuff? It is really helpful for debugging derive with composite keys.

**Comment by udoprog** [comment]
> Understood on the examples, but can you restore the one you deleted (`map_feature`) and just remove the feature stuff? It is really helpful for debugging derive with composite keys.

Ah. Not sure I want to have a permanent example for this since they're meant to be demonstrative uses of the library. :thinking: What if we keep it as a test? Then you can run it with `cargo build --test <name>`?

**Comment by udoprog** [comment]
The `empty` example should probably also be moved*!

**Comment by udoprog** [comment]
All right, with this you can now check all expansions with:

```
cargo install cargo-expand # if you don't already have it
cargo expand --test map_feature
```

Be warned: it's VERY noisy. :)

Does that work for you?

**Comment by pitaj** [comment]
That works, thanks

**Pull request merged**